### PR TITLE
[FLINK-3580] [table] Reintroduce Date/Time and implement scalar functions for it

### PR DESCRIPTION
**Pull request by twalthr**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

This PR implements basic Date/Time/Timestamp support for the Table API. Not all scalar functions are supported so far. It also solves FLINK-3152. I will add some documentation later.

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/typeutils/TypeCoercion.scala [hunk 1]
```diff
@@ -85,6 +85,9 @@ object TypeCoercion {
     case (STRING_TYPE_INFO, _: NumericTypeInfo[_]) => true
     case (STRING_TYPE_INFO, BOOLEAN_TYPE_INFO) => true
     case (STRING_TYPE_INFO, BIG_DEC_TYPE_INFO) => true
+    case (STRING_TYPE_INFO, SqlTimeTypeInfo.DATE) => true
+    case (STRING_TYPE_INFO, SqlTimeTypeInfo.TIME) => true
+    case (STRING_TYPE_INFO, SqlTimeTypeInfo.TIMESTAMP) => true
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/typeutils/TypeCoercion.scala]: Allow casting from LONG to TIMESTAMP?

**Comment by fhueske** [comment]
Hi @twalthr, thanks for the PR. I skimmed the changes and everything looked fine. I have two suggestions though. 

1. Can we add a casts from `Long` to `Timestamp` and back? 
2. Can you add a Date, Time, and Timestamp type an existing ITCase for Scala SQL and Table API to check that these work fine?

Thanks, Fabian

**Comment by twalthr** [comment]
@fhueske I have updated the PR. If there are no objections, I would merge it soon. So that it will be contained in the next release.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/sql/ExpressionsITCase.scala [hunk 2]
```diff
@@ -120,4 +122,26 @@ class ExpressionsITCase(
     TestBaseUtils.compareResultAsText(results.asJava, expected)
   }
 
+  @Test
+  def testAdvancedDataTypes(): Unit = {
```
Comment (aljoscha) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/sql/ExpressionsITCase.scala]: this should probably be `testAdvancedDateTypes`

File: docs/apis/table.md [hunk 3]
```diff
@@ -775,6 +775,8 @@ column names and function names follow Java identifier syntax. Expressions speci
 
 If working with exact numeric values or large decimals is required, the Table API also supports Java's BigDecimal type. In the Scala Table API decimals can be defined by `BigDecimal("123456")` and in Java by appending a "p" for precise e.g. `123456p`.
```
Comment (aljoscha) [inline docs/apis/table.md]: Is the `p` suffix a standard suffix for BigDecimal?

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/sql/ExpressionsITCase.scala (hunk 2 above)
Comment (twalthr) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/sql/ExpressionsITCase.scala]: This is meant to be the unit test for all future advanced data types. It has the same name as the test in `ExpressionITCase` of Scala Table API.

File: docs/apis/table.md (hunk 3 above)
Comment (twalthr) [inline docs/apis/table.md]: No, there is no standard suffix for BigDecimals and `d` is already the suffix for `double`.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/sql/ExpressionsITCase.scala (hunk 2 above)
Comment (aljoscha) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/sql/ExpressionsITCase.scala]: Ah, I see, thanks.

**Comment by twalthr** [comment]
Merging...

**Pull request closed without merging**